src/commander.js

```diff
@@ -72,30 +68,96 @@ define([
         cmd.input.val("").removeClass("alert-danger");
     }
 
-    function setupAutocomplete(cmd, choices) {
-        var input = cmd.input;
-        input.atwho({
-            at: "",
-            data: choices,
-            limit: 50,
-            maxLen: Infinity,
-            suffix: " ",
-            tabSelectsMatch: true,
-            callbacks: {
-                // match first argument only
-                matcher: function (flag, subtext) { return subtext; }
-            }
-        });
+    function setupAutocomplete(cmd) {
+        if (cmd.atwhoConfig) {
+            return;
+        }
+        var typeNames = _.pluck(fn.getQuestionMap(cmd.vellum), "typeName"),
+            positions = ["after", "before", "in", "first in"],
+            questionRe = /[#\/][^\s]+/;
+        cmd.atwhoConfig = getAtwhoConfig([
+            [literals(typeNames), literals(positions), {regexp: questionRe.source}],
+            [{regexp: questionRe.source}],
+        ]);
+        cmd.input.atwho(cmd.atwhoConfig);
         atwho.autocomplete(
-            input,
+            cmd.input,
             {form: cmd.vellum.data.core.form, on: _.identity},  // fake mug
             {useHashtags: true, tabSelectsMatch: true}          // options
         );
-        input.on("inserted.atwho", function (event, item) {
+        cmd.input.on("inserted.atwho", function (event) {
             event.preventDefault();
         });
     }
 
+    function literals(names) {
+        var source = "(" + _.map(names, RegExp.escape).join("|") + ")";
+        return {regexp: source, names: names};
+    }
+
+    function getAtwhoConfig(commandForms) {
```

Comment:
A summarizing comment here would be helpful, there's a lot going on here (programmatically generated regexes!).